NIFI-13048 - Add build info, bundle size and NiFi API version for bundles in NiFi Registry

pvillard31 opened the pull request:
# Summary

[NIFI-13048](https://issues.apache.org/jira/browse/NIFI-13048) - Add build info, bundle size and NiFi API version for bundles in NiFi Registry. It also fixes the action menu to not hardcode 'Flow' as the action menu can be used for both bundles and flows.

Here is a screenshot of NiFi Registry with bundles before the change:

![Screenshot 2024-04-15 at 15 00 02](https://github.com/apache/nifi/assets/11541012/8ac2cd9e-9553-4824-97da-555ffb0a2fbe)

Here is a screenshot of NiFi Registry with bundles after the change:

![Screenshot 2024-04-15 at 17 44 26](https://github.com/apache/nifi/assets/11541012/408e680a-a007-494e-af18-61b47aec0896)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

exceptionfactory commented on nifi-registry/nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/registry/nf-registry-grid-list-viewer.html:
```diff
@@ -121,6 +121,38 @@
+                                            <div fxLayout="row" class="mat-body-3">
+                                                <p>
+                                                Build Information<br />
+                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>NiFi API version</b>: {{snapshotMeta.systemApiVersion}}<br />
```
Using non-breaking spaces for formatting is not ideal, so I recommend using styled `div` tags instead.

exceptionfactory commented on nifi-registry/nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/registry/nf-registry-grid-list-viewer.html:
```diff
@@ -121,6 +121,38 @@
+                                                <p>
+                                                Build Information<br />
+                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>NiFi API version</b>: {{snapshotMeta.systemApiVersion}}<br />
+                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>Build branch</b>: {{snapshotMeta.buildInfo.buildBranch}}<br />
```
Recommend capitalizing the second word in these lines:
```suggestion
                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>Build Branch</b>: {{snapshotMeta.buildInfo.buildBranch}}<br />
```

exceptionfactory commented on nifi-registry/nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/registry/nf-registry-grid-list-viewer.html:
```diff
@@ -121,6 +121,38 @@
+                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>Build tag</b>: {{snapshotMeta.buildInfo.buildTag}}<br />
+                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>Build tool</b>: {{snapshotMeta.buildInfo.buildTool}}<br />
+                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>Built by</b>: {{snapshotMeta.buildInfo.builtBy}}<br />
+                                                &nbsp;&nbsp;&nbsp;&nbsp;<b>Built on</b>: {{(snapshotMeta.buildInfo.built/1000 | amFromUnix) | amDateFormat:'MMM-DD-YYYY'}} at {{(snapshotMeta.buildInfo.built/1000 | amFromUnix) | amDateFormat:'h:mm A'}}
```
Is the `MMM-DD-YYYY` format common elsewhere in the application? I recommend using the ISO 8601 format of `YYYY-MM-DD`.

pvillard31 replied:
It's also used for the date at which the version has been added in the NiFi Registry, but agree that using the other format is better

exceptionfactory reviewed (commented):
Thanks for proposing this improvement @pvillard31. Having the build information visible is useful. I noted a couple recommendations on formatting, but the general approach seems straightforward.

pvillard31 commented:
Thanks for the review @exceptionfactory - I believe I addressed your comments.

![Screenshot 2024-04-16 at 16 08 33](https://github.com/apache/nifi/assets/11541012/b25c5a22-39f4-4d9b-be2e-12b02d0a1303)

exceptionfactory commented on nifi-registry/nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/registry/nf-registry-grid-list-viewer.html:
```diff
@@ -115,7 +115,39 @@
+                                        <div fxLayout="column" fxLayoutAlign="space-between stretch">
+                                            <div fxLayout="row" class="mat-body-3">
+                                                <div>Build Information:</div>
+                                                <div style="margin-left: 14px; margin-bottom: 5px;">
```
The inline style should be replaced with an externalized class.

exceptionfactory commented on nifi-registry/nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/registry/nf-registry-grid-list-viewer.html:
```diff
@@ -115,7 +115,39 @@
+                                            <div fxLayout="row" class="mat-body-3">
+                                                <div>Build Information:</div>
+                                                <div style="margin-left: 14px; margin-bottom: 5px;">
+                                                    <b>NiFi API Version</b>: {{snapshotMeta.systemApiVersion}}<br />
```
Instead of using `<br/>`, I believe it would be better to use one `<div>` per row.

exceptionfactory reviewed (changes requested):
Thanks for the updates @pvillard31, the date format changes looks good.

The styling approach is better, but the inline style should be replaced with a CSS class. I also recommend using a `<div>` per row instead of the `<br/>`.

pvillard31 commented:
Tried many options and finally went for ul/li which seems to be the cleanest approach compared to div and CSS.

![Screenshot 2024-04-16 at 18 10 15](https://github.com/apache/nifi/assets/11541012/b4b27394-b35a-4906-8107-c08adb6dc160)

exceptionfactory commented:
> Tried many options and finally went for ul/li which seems to be the cleanest approach compared to div and CSS.

Thanks, the bullet list approach looks good.

exceptionfactory reviewed (approved):
Thanks for making the adjustments @pvillard31, the latest version looks good. +1 merging
